src/Umbraco.Infrastructure/DependencyInjection/UmbracoBuilder.Uniques.cs

```diff
@@ -109,18 +109,53 @@ public static void SetShortStringHelper(this IUmbracoBuilder builder, IShortStri
+        /// Sets the filesystem used by the MediaFileManager
         /// </summary>
         /// <param name="builder">A builder.</param>
-        /// <param name="filesystemFactory">A filesystem factory.</param>
+        /// <param name="filesystemFactory">Factory method to create an IFileSystem implementation used in the MediaFileManager</param>
+        public static void SetMediaFileSystem(this IUmbracoBuilder builder,
+            Func<IServiceProvider, IFileSystem> filesystemFactory) => builder.Services.AddSingleton(
+            provider =>
+            {
+                IFileSystem filesystem = filesystemFactory(provider);
+                // We need to use the Filesystems to create a shadow wrapper,
+                // because shadow wrapper requires the IsScoped delegate from the FileSystems.
+                // This is used by the scope provider when taking control of the filesystems.
+                FileSystems fileSystems = provider.GetRequiredService<FileSystems>();
+                IFileSystem shadow = fileSystems.CreateShadowWrapper(filesystem, "media");
+
+                return provider.CreateInstance<MediaFileManager>(shadow);
+            });
+
+        /// <summary>
+        /// Register FileSystems with a specific IFileSystem for stylesheets
+        /// </summary>
         /// <remarks>
-        /// Using this helper will ensure that your IFileSystem implementation is wrapped by the ShadowWrapper
+        /// Be careful when using this, the root path and root url must be correct for this to work.
         /// </remarks>
-        public static void SetMediaFileSystem(this IUmbracoBuilder builder, Func<IServiceProvider, IFileSystem> filesystemFactory)
-            => builder.Services.AddUnique<IMediaFileSystem>(factory =>
+        /// <param name="builder">A builder.</param>
+        /// <param name="stylesheetFactory">Factory method to create an IFileSystem implementation used for stylesheets.</param>
+        /// <exception cref="InvalidOperationException">Throws exception if full path can't be resolved successfully.</exception>
+        public static void AddFileSystems(this IUmbracoBuilder builder,
+            Func<IServiceProvider, IFileSystem> stylesheetFactory) => builder.Services.AddUnique(
+            provider =>
             {
-                var fileSystems = factory.GetRequiredService<FileSystems>();
-                return fileSystems.GetFileSystem<MediaFileSystem>(filesystemFactory(factory));
+                IFileSystem stylesheetFileSystem = stylesheetFactory(provider);
+                // Verify that _rootUrl/_rootPath is correct
+                // We have to do this because there's a tight coupling
+                // to the VirtualPath we get with CodeFileDisplay from the frontend.
+                try
+                {
+                    var rootPath = stylesheetFileSystem.GetFullPath("/css/");
+                }
+                catch (UnauthorizedAccessException exception)
```

Comment:
Should be `InvalidOperationException`

Reply:
Not sure I understand this? GetFullPath throws an UnauthorizedAccessException if the path is not set up correctly, but I suppose we should re throw an UnauthorizedAccessException then but with a different message added?